contracts/staking/Staking.sol

```diff
@@ -0,0 +1,193 @@
+    }
+
+    /**
+     * @notice Claims the specified amount of rewards.
+     * @param amount_ The amount of rewards to claim.
+     */
+    function claim(uint256 amount_) public stakingStarted {
+        uint256 owed = userDistribution(msg.sender).owedValue;
+        require(amount_ <= owed, "Staking: insufficient amount");
+
+        _distributeValue(msg.sender, amount_);
+    }
+
+    /**
+     * @notice Returns the shares token.
+     * @return The address of the shares token contract.
+     */
+    function sharesToken() public view returns (address) {
+        return _sharesToken;
+    }
+
+    /**
+     * @notice Returns the rewards token.
+     * @return The address of the rewards token contract.
+     */
+    function rewardsToken() public view returns (address) {
+        return _rewardsToken;
+    }
+
+    /**
+     * @notice Returns the staking start time.
+     * @return The timestamp when staking started.
+     */
+    function stakingStartTime() public view returns (uint256) {
+        return _stakingStartTime;
+    }
+
+    /**
+     * @notice Returns the rate of rewards distribution.
+     * @return The rate of rewards distribution per second.
+     */
+    function rate() public view returns (uint256) {
+        return _rate;
+    }
+
+    /**
+     * @notice Sets the staking start time.
+     * @param stakingStartTime_ The timestamp when staking will start.
+     */
+    function _setStakingStartTime(uint256 stakingStartTime_) internal {
+        _stakingStartTime = stakingStartTime_;
+    }
+
+    /**
+     * @notice Sets the rate of rewards distribution per second.
+     * @param newRate_ The new rate of rewards distribution.
+     */
+    function _setRate(uint256 newRate_) internal {
+        _update(address(0));
+
+        _rate = newRate_;
+    }
+
+    /**
+     * @notice Hook function that is called after shares have been added to a user's distribution.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of shares added.
+     */
+    function _afterAddShares(address user_, uint256 amount_) internal virtual override {
+        SafeERC20.safeTransferFrom(IERC20(_sharesToken), user_, address(this), amount_);
+    }
+
+    /**
+     * @notice Hook function that is called after shares have been removed from a user's distribution.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of shares removed.
+     */
+    function _afterRemoveShares(address user_, uint256 amount_) internal virtual override {
+        SafeERC20.safeTransfer(IERC20(_sharesToken), user_, amount_);
+    }
+
+    /**
+     * @notice Hook function that is called after value has been distributed to a user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value distributed.
+     */
+    function _afterDistributeValue(address user_, uint256 amount_) internal virtual override {
+        SafeERC20.safeTransfer(IERC20(_rewardsToken), user_, amount_);
```

Comment:
Let's make the code more attractive with `using` directives. Then you will be able to call IERC20(_rewardsToken).safeTransfer directly